src/algebra/star/subalgebra.lean

```diff
@@ -225,3 +250,341 @@ set.centralizer_subset h
+  (Hmul_left : ∀ (x₁ x₂ y : A), p x₁ y → p x₂ y → p (x₁ * x₂) y)
+  (Hmul_right : ∀ (x y₁ y₂ : A), p x y₁ → p x y₂ → p x (y₁ * y₂))
+  (Hstar : ∀ (x y : A), p x y → p (star x) (star y))
+  (Hstar_left : ∀ (x y : A), p x y → p (star x) y)
+  (Hstar_right : ∀ (x y : A), p x y → p x (star y)) : p a b :=
+begin
+  refine algebra.adjoin_induction₂ ha hb (λ x hx y hy, _) Halg (λ r x hx, _) (λ r x hx, _)
+    Hadd_left Hadd_right Hmul_left Hmul_right,
+  { cases hx; cases hy,
+    exacts [Hs x hx y hy, star_star y ▸ Hstar_right _ _ (Hs _ hx _ hy),
+      star_star x ▸ Hstar_left _ _ (Hs _ hx _ hy),
+      star_star x ▸ star_star y ▸ Hstar _ _ (Hs _ hx _ hy)] },
+  { cases hx, exacts [Halg_left _ _ hx, star_star x ▸ Hstar_right _ _ (Halg_left r _ hx)] },
+  { cases hx, exacts [Halg_right _ _ hx, star_star x ▸ Hstar_left _ _ (Halg_right r _ hx)] },
+end
+
+/-- The difference with `star_subalgebra.adjoin_induction` is that this acts on the subtype. -/
+lemma adjoin_induction' {s : set A} {p : adjoin R s → Prop} (a : adjoin R s)
+  (Hs : ∀ x (h : x ∈ s), p ⟨x, subset_adjoin R s h⟩)
+  (Halg : ∀ r, p (algebra_map R _ r)) (Hadd : ∀ x y, p x → p y → p (x + y))
+  (Hmul : ∀ x y, p x → p y → p (x * y)) (Hstar : ∀ x, p x → p (star x)) : p a :=
+subtype.rec_on a $ λ b hb,
+begin
+  refine exists.elim _ (λ (hb : b ∈ adjoin R s) (hc : p ⟨b, hb⟩), hc),
+  apply adjoin_induction R hb,
+  exacts [λ x hx, ⟨subset_adjoin R s hx, Hs x hx⟩,
+    λ r, ⟨star_subalgebra.algebra_map_mem _ r, Halg r⟩,
+    (λ x y hx hy, exists.elim hx $ λ hx' hx,
+      exists.elim hy $ λ hy' hy, ⟨add_mem hx' hy', Hadd _ _ hx hy⟩),
+    (λ x y hx hy, exists.elim hx $ λ hx' hx,
+      exists.elim hy $ λ hy' hy, ⟨mul_mem hx' hy', Hmul _ _ hx hy⟩),
+    λ x hx, exists.elim hx (λ hx' hx, ⟨star_mem hx', Hstar _ hx⟩)]
+end
+
+/-- If all elements of `s : set A` commute pairwise and also commute pairwise with elements of
+`star s`, then `star_subalgebra.adjoin R s` is commutative. See note [reducible non-instances]. -/
+@[reducible]
+def adjoin_comm_semiring_of_comm {s : set A}
+  (hcomm : ∀ (a : A), a ∈ s → ∀ (b : A), b ∈ s → a * b = b * a)
+  (hcomm_star : ∀ (a : A), a ∈ s → ∀ (b : A), b ∈ s → a * star b = star b * a) :
+comm_semiring (adjoin R s) :=
+{ mul_comm :=
+  begin
+    rintro ⟨x, hx⟩ ⟨y, hy⟩,
+    ext,
+    simp only [set_like.coe_mk, mul_mem_class.mk_mul_mk],
+    rw [←mem_to_subalgebra, adjoin_to_subalgebra] at hx hy,
+    letI : comm_semiring (algebra.adjoin R (s ∪ star s)) := algebra.adjoin_comm_semiring_of_comm R
+    begin
+      intros a ha b hb,
+      cases ha; cases hb,
+      exacts [hcomm _ ha _ hb, star_star b ▸ hcomm_star _ ha _ hb,
+        star_star a ▸ (hcomm_star _ hb _ ha).symm,
+        by simpa only [star_mul, star_star] using congr_arg star (hcomm _ hb _ ha)],
+    end,
+    exact congr_arg coe (mul_comm (⟨x, hx⟩ : algebra.adjoin R (s ∪ star s)) ⟨y, hy⟩),
+  end,
+  ..(adjoin R s).to_subalgebra.to_semiring }
```

Comment:
Easier, yes. I was doing it this way to avoid potential diamonds, or at least preserve definitional unfolding, but maybe that concern is unfounded in this case. I'll check into it later.